### PR TITLE
Final PR: Umdtask144 fall2025 fake news detection

### PR DESCRIPTION
**Pull request by dhanushv2000**
## Summary of Changes

This PR implements a **production-ready fake news detection system** using a fine-tuned **BERT transformer model**, deployed via a **Model Context Protocol (MCP) server** with REST APIs, a web interface, and Docker-based deployment. The project moves beyond a notebook-only ML workflow and delivers a **fully accessible AI service** for real-time and batch fake news classification with confidence scores.

---

### Core Implementation
- **`mcp_server.py`** – Flask-based MCP REST server exposing standardized endpoints for prediction, model management, and health checks  
- **`mcp_server_class.py`** – `BERTMCPServer` implementation handling model lifecycle, inference, logging, and statistics  
- **`bert_utils.py`** – Utilities for BERT loading, tokenization, and inference  
- **`preprocessing.py`** – Text cleaning, normalization, and preprocessing utilities  

### Notebooks
- **`mcp_fake_news.example.ipynb`** – End-to-end walkthrough (dataset → preprocessing → modeling → evaluation → MCP deployment)  
- **`BERT_training.ipynb`** – Fine-tuning BERT for fake news detection with cross-validation and metric evaluation  
- **`Data_Preprocessing.ipynb`** – Detailed preprocessing and exploratory analysis  

### Documentation
- **`README.md`** – Project overview, architecture, dataset description, performance metrics, and usage instructions  
- **`Example.md`** – Training methodology and ML pipeline explanation  
- **`API.md`** – MCP REST API documentation with request/response examples  

### Model Artifacts
- **`models/bert_fake_news/`** – MCP-managed trained BERT model (`pytorch_model.bin`, `config.json`, `tokenizer_config.json`)  

### Deployment & Configuration
- **`Dockerfile.mcp`** – Production-ready Docker image for MCP server  
- **`docker-compose.mcp.yml`** – MCP server orchestration  
- **`docker_manage.sh`** – Interactive Docker management script  
- **`docker_build.sh`**, **`docker_run.sh`**, **`docker_stop.sh`**, **`docker_logs.sh`**, **`docker_clean.sh`** – Deployment utilities  
- **`requirements.txt`** – Python dependencies and pinned versions  

### Web Interface
- **`templates/index.html`** – Interactive web UI for real-time fake news prediction  

### Data
- **`data/true.csv`** – ~21,000 real news articles (Reuters and professional sources)  
- **`data/fake.csv`** – ~23,000 fake news articles (known misinformation sources)  

---

### Dataset
- **Total Articles:** ~44,000  
- **Class Distribution:** Fake (52%), Real (48%)  
- **Sources:** Reuters and verified news outlets (real), misinformation websites (fake)  

### Model Performance
- **Accuracy:** 84.74%  
- **Precision:** 82.13%  
- **Recall:** 87.97%  
- **F1-Score:** 84.95%  
- **ROC-AUC:** 0.936  

*The model generalizes well to unseen data and is suitable as a first-pass filter for fake news detection.*

---

## Key Features
- Fine-tuned **BERT (bert-base-uncased)** for contextual language understanding  
- **MCP-compliant REST API** for standardized model access  
- **Real-time & batch prediction** support  
- **Confidence scores** and inference timing  
- **Web UI** for non-technical users  
- **Dockerized deployment** for reproducibility and portability  
- **Extensible MCP architecture** supporting future models and versioning  

---

### Docker (Recommended)
```bash
./docker_manage.sh
# Choose option 8 for full setup
```

- Web UI: http://localhost:9090/  
- MCP REST API available on port **9090**

### Local Development
```bash
pip install -r requirements.txt
python mcp_server.py
```

### Notebooks
```bash
jupyter lab mcp_fake_news.example.ipynb
jupyter lab BERT_training.ipynb
```

---

## Notes & Limitations
- Articles are truncated to manage long inputs  
- BERT operates on a limited context window  
- Model detects **language patterns**, not factual truth  
- Performance depends on training data distribution  

---

## Future Improvements
- Multi-model support via MCP (BERT, DistilBERT, ensembles)  
- Model versioning and A/B testing  
- Explainability endpoints (SHAP / LIME)  
- Domain-specific fake news models  
- Integration with external fact-checking APIs  

---

All files follow **MSML610 course project guidelines**, use a clean directory structure, and demonstrate a **complete ML system from training to production deployment**.

### CODE REVIEW

**Comment by aver81** [comment]
@dhanushv2000 I can't find API.md and example.md files. Please make sure they are also present

**Comment by aver81** [comment]
@dhanushv2000 The content for the API.md is incorrect. the API files should only focus on the tool itself, what it is, how to use and some code to show it usage. For example, the API jupyter notebook should not be connected in any way to your project. all project implementation must be contained in the example jupyter notebook. Similarly, for API.md, it should talk about the tool itself. All project related implementation that involves the tool should go in example.md
